source/common/config/grpc_mux_impl.cc

```diff
@@ -336,6 +337,15 @@ void GrpcMuxImpl::expiryCallback(absl::string_view type_url,
 
     watch->callbacks_.onConfigUpdate({}, found_resources_for_watch, "");
   }
+
+  // Now that we've forgotten about the resource, enqueue a new request with an empty version to
```

Comment:
@snowp from a design perspective, would it be desirable to try fetching a refresh before TTL expires? It seems there are two scenarios.. imagine we have a 30 minute TTL.
1. The control plane is intelligent and push a refresh at let's say 25 minutes for another 30 minute TTL. There is no disruption of data plane traffic.
2. The control plane is not intelligent. Then the client expires at 30 minutes, traffic is disrupted and then a new resource is fetched for another 30 minutes.

We could avoid the disruption in (2) by being a bit smarter client side.

Reply:
The nice thing about making the client smarter here is that the server now becomes basically unaware of TTL expiration (beyond setting the flag), just responding to requests. I'm not sure if it's actually worth the complexity in the client though, and it also might limit the usage of TTLs, for example in cases where the server know that it wants some configuration to be active for N minutes so it sets the TTL to N minutes and never wants to refresh it. I think possibly some configuration for TTL pre-emptive fetching would be useful, but I'd probably wait to hear from more people that this is something people are interested in first. 

I think the motivation for this PR was rather to provide some mechanism for the client to say that it has expired resources, but the more I think about this the more I think we need to have the server proactively do a "just in case" push of the TTL'd resources, unless include the remaining TTL in the request (which seems overkill). The server needs to do this push in order to reset all the TTL timers so that it has a sense of how much time is remaining on the timers, otherwise it has no idea when it needs to push a heartbeat update to refresh the TTL.

Reply:
Yeah, I think in the cases of perpetual heartbeat, the server should do the work. What I'm wondering is if it's ever useful to provide a "just expired" mechanism? Basically, by time you send this, it's too late to do anything useful, you're already dropping traffic and so you're in a state you don't want to be in.